rust/lance-core/src/utils/backoff.rs

```diff
@@ -73,6 +77,64 @@ impl Backoff {
     }
 }
 
+/// SlotBackoff is a backoff strategy that uses a slot-based approach.
+///
+/// This is for when the cause of the failure is concurrency itself and that
+/// the attempts take about the same amount of time.
+///
+/// Say you have N attempts to do something. We don't know there are N ahead of
+/// time. We start guessing with 4 slots:
+///
+/// | 1, 2, 3 | 4, 5, 6 | 7, 8, 9 | 10 |
+///
+/// Each slot can have one success, so we can eliminate 3, 6, 9, and 10. In the
+/// next round, we will use twice as many slots (8):
+///
+/// | 1 | 2 | 4 | 5 | 7 | 8 | ... |
+///
+/// Optimally, that should be 16 total attempts. For a 1s unit, the retry times
+/// are:
+/// * Round 1: 0 - 3s
+/// * Round 2: 0 - 7s
+/// * Round 3: 0 - 15s
+/// * ...
```

Comment:
I find the explanation a little hard to follow.  I get the idea of slots but I have a hard time mapping that back to a backoff time.  I think the key here is there is some "unit duration" and each slot represents one unit of time.  So when we need to calculate backoff we randomly pick a slot and then space out the slots so they are "unit duration" apart from each other?